pilot/pkg/networking/apigen/apigen_test.go

```diff
@@ -0,0 +1,193 @@
+		{
+			Address:         "127.0.0.1",
+			EndpointPort:    uint32(14056),
+			ServicePortName: "http-main",
+		},
+	})
+	se := collections.IstioNetworkingV1Alpha3Serviceentries.Resource()
+	store := ds.MemoryConfigStore
+
+	store.Create(model.Config{
+		ConfigMeta: model.ConfigMeta{
+			Type:      se.Kind(),
+			Group:     se.Group(),
+			Version:   se.Version(),
+			Name:      "fortio",
+			Namespace: "fortio",
+		},
+		Spec: &networking.ServiceEntry{
+			Hosts: []string{
+				"fortio.fortio.svc",
+				"fortio.fortio.svc.cluster.local",
+			},
+			Addresses: []string{"1.2.3.4"},
+
+			Ports: []*networking.Port{
+				{Number: 14056, Name: "grpc-insecure", Protocol: "http"},
+			},
+
+			Endpoints: []*networking.WorkloadEntry{
+				{
+					Address: "127.0.0.1",
+					Ports:   map[string]uint32{"grpc-insecure": 8080},
+				},
+			},
+			Location:   networking.ServiceEntry_MESH_EXTERNAL,
+			Resolution: networking.ServiceEntry_STATIC,
+		},
+	})
+
+	env := ds.DiscoveryServer.Env
+	if err := env.PushContext.InitContext(env, env.PushContext, nil); err != nil {
+		t.Fatal(err)
+	}
+	ds.DiscoveryServer.UpdateServiceShards(env.PushContext)
+
+	err := ds.StartGRPC(grpcAddr)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer ds.GRPCListener.Close()
+
+	os.Setenv("GRPC_XDS_BOOTSTRAP", "testdata/xds_bootstrap.json")
+
+	t.Run("gRPC-resolve", func(t *testing.T) {
+		rb := resolver.Get("xds-experimental")
+		ch := make(chan resolver.State)
+		_, err := rb.Build(resolver.Target{Endpoint: istiodSvcAddr},
+			&testClientConn{ch: ch}, resolver.BuildOptions{})
+
+		if err != nil {
+			t.Fatal("Failed to resolve XDS ", err)
+		}
+		tm := time.After(10 * time.Second)
+		select {
+		case s := <-ch:
+			log.Println("Got state ", s)
+		// TODO: timeout
+		case <-tm:
+			t.Error("Didn't resolve")
+		}
+	})
+
+	t.Run("gRPC-cdslb", func(t *testing.T) {
+		rb := balancer.Get("eds_experimental")
+		//ch := make(chan resolver.State)
+		b := rb.Build(&testLBClientConn{}, balancer.BuildOptions{})
+
+		defer b.Close()
+
+	})
+
+	t.Run("gRPC-dial", func(t *testing.T) {
+		conn, err := grpc.Dial("xds-experimental:///istiod.istio-system.svc.cluster.local:14056", grpc.WithInsecure())
+		if err != nil {
+			t.Fatal("XDS gRPC", err)
+		}
+
+		defer conn.Close()
+		xds := ads.NewAggregatedDiscoveryServiceClient(conn)
+
+		s, err := xds.StreamAggregatedResources(context.Background())
+		if err != nil {
+			t.Fatal(err)
+		}
+		s.Send(&xdsapi.DiscoveryRequest{})
+
+	})
+
+}
+
+type testLBClientConn struct {
+	balancer.ClientConn
+}
+
+// From xds_resolver_test
+// testClientConn is a fake implemetation of resolver.ClientConn. All is does
+// is to store the state received from the resolver locally and signal that
+// event through a channel.
+type testClientConn struct {
+	resolver.ClientConn
+	ch chan resolver.State
+}
+
+func (t *testClientConn) UpdateState(s resolver.State) {
+	log.Println("UPDATE STATE ", s)
```

Comment:
remove this or make it `test.log`

Reply:
I'll remove, was for debugging to make sure it's called.

Reply:
And to find what they are sending, I was curious.